### ISSUE TITLE
Queries language: passing adapter specific parameters

### ISSUE THREAD

**Issue by dmarcelino** [balderdashy/waterline#780]
Hi! Generally speaking it's a good idea for queries to be adapter agnostic to maximize code re-usability. But there are times where a simple adapter specific parameter can save a lot of work, e.g., preventing the developer from having to bypass waterline altogether and make a direct call to the DB driver. I've noticed that waterline actually lets custom parameters pass through to the adapter layer, which is nice, but I wonder if there are or should be any kind of convention for these custom parameters.

In waterline-orientdb, I'm now processing a parameter with name `_orientdb` which allow queries like:
``` javascript
Person.findOne({ where: { name: 'Luca' }, _orientdb: { fetchPlanLevel: 2 } })
```
Working example [here](https://github.com/appscot/waterline-orientdb/blob/master/test/integration-orientdb/tests/associations/manyThrough.find.js#L169). I'm wondering, is this a good way of doing it? Is it future proof? Please let me know if waterline has any conventions for adapter specific parameters or lets discuss if there should be. Thanks!

**Comment by particlebanana** [comment]
This is something that I'll open up to the community. Right now it's a "bug" in that it just happens to fall through. Lots of people use this fact to pass in options to mongo so we should probably look at making a standard for it.

**Comment by dmarcelino** [comment]
Thanks @particlebanana, given I raised the issue I'll be the first to throw in a suggestion: `_adapter`. Example:
```javascript
Person.findOne({ where: { name: 'walter' }, _adapter: { someAdapterOption: 'option' } })
```

Being named "adapter" should suffice to indicate to the user that these are adapter specific options. The underscore in front of it should suggest caution as, generally speaking, users shouldn't be using adapter specific options as that may break compatibility with other adapters.

I also suggest we clearly point out in the [docs](https://github.com/balderdashy/waterline-docs/blob/master/query-language.md#query-options) that the use of adapter options is discouraged for the above reasons.

I don't feel too strong about the specific name we choose for this as long as we can agree on one so adapters can start adopting it.

**Comment by randallmeeker** [comment]
By adapter specific could we mean connection specific? Have the underscored variable reference a particular connection provided it is being used. I have had situations where I have moved my database between either two different servers (MySQL to MySQL) or two different platforms (MySQL to Mongo) and in other frameworks I could list connection specific properties based on the connection I was using.

**Comment by particlebanana** [comment]
@randallmeeker I think the tenant issue would solve some of what you are asking for: https://github.com/balderdashy/waterline/pull/787. This I see as something to make Waterline work with more of the non-traditional databases like couchdb and orientdb. Either way between this and the tenant support you should be able to get anything you need down to the adapter level.

**Comment by randallmeeker** [comment]
@particlebanana I would think using the tenant approach would be a hack for what that is proposed for, when this seem intended for adapter specific language. Granted if both of these were implemented then there is pretty much no limit.

**Comment by atiertant** [comment]
@dmarcelino adapter functionality is a large problem,
adapter could not synchronize all functionalities and some are adapter specific like you're saying.
to be useable in waterline, adapters should declare them in npm package.json to let adapter-test testing them like interface(see https://github.com/balderdashy/waterline-adapter-tests/pull/64).
adapter should export prefixed function and criteria operator.
but the problem with interface mecanism is that their number will incease fastly,because adding a functionnality result in adding an interface (see https://github.com/balderdashy/waterline-adapter-tests/issues/12#issuecomment-33417854)

i would like prefixed _sql functions and operator that could be implement in all sql adapter.
sql interface only have a "query" fonction,in this case,it could be wrote _sql-query or --sql-query,so addind and other function to an sql adapter does'nt need create an other interface.

**Comment by dmarcelino** [comment]
Hey @atiertant, I'm quite familiar with balderdashy/waterline-adapter-tests#64, I wrote it! :grinning:  Interfaces (and one day "features") implement functionality approved and prescribed by waterline core maintainers and (hopefully) tested in waterline-adapter-tests to ensure compliance. We could even call that kind of functionality *standard*. Because that kind of functionality always tries to fit more than one adapter and since it requires waterline-adapter-tests changes, they will always take more time to go live. What I proposed here is slightly different...

My proposal in this issue goes in the direction of opening a door for non-standard / non-approved adapter-specific functionality at the `criteria` level (I'm not referring to custom methods since adapters are already free to add those). So if a given adapter supports a funky criteria option such as:
``` javascript
Person.findOne({ where: { name: 'walter' }, _adapter: { doTheFunkyThing: true } })
```
There should be a convention for a user to pass that option to the adapter, in my opinion.

Currently I'm using `_orientdb` for such sails-orientdb specific parameters (check [manyThrough.find.js#L182](https://github.com/appscot/sails-orientdb/blob/master/test/integration-orientdb/tests/associations/manyThrough.find.js#L182)) but I have no guarantee that one day a change in waterline core won't break this for me...

**Comment by atiertant** [comment]
@dmarcelino i just tried to explain that i think waterline will need a strong feature management because adapter are all differents and all function couldn't be wrote easily in some adapter.
maybe it could be interresting to have only one notation for interface,adapter features,(and one day "features")...like browser and css : in first time  there is a  prefix browser specific -moz- , -webkit- , -o- , -ms- ,.... and one day this could become standard and strip prefix

**Comment by dmarcelino** [comment]
I see, yes, I generally agree with what you're saying.

**Comment by sailsbot** [comment]
Thanks for posting, @dmarcelino.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
